pkg/dumper/generic/sql.go

```diff
@@ -0,0 +1,150 @@
+package generic
+
+import (
+	"sync"
+
+	"github.com/hellofresh/klepto/pkg/config"
+	"github.com/hellofresh/klepto/pkg/database"
+	"github.com/hellofresh/klepto/pkg/dumper"
+	"github.com/hellofresh/klepto/pkg/reader"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+type (
+	sqlDumper struct {
+		SqlEngine
+
+		reader reader.Reader
+	}
+
+	SqlEngine interface {
+		DumpStructure(sql string) error
+
+		DumpTable(tableName string, rowChan <-chan database.Row) error
+
+		// Close closes the dumper resources and releases them.
+		Close() error
+	}
+
+	SqlEngineAdvanced interface {
+		PreDumpTables([]string) error
+		PostDumpTables([]string) error
+	}
+)
+
+func NewSqlDumper(rdr reader.Reader, engine SqlEngine) dumper.Dumper {
+	return &sqlDumper{
+		SqlEngine: engine,
+		reader:    rdr,
+	}
+}
+
+func (p *sqlDumper) Dump(done chan<- struct{}, configTables config.Tables) error {
+	if err := p.readAndDumpStructure(); err != nil {
+		return err
+	}
+
+	return p.readAndDumpTables(done, configTables)
+}
+
+func (p *sqlDumper) readAndDumpStructure() error {
+	log.Debug("Dumping structure...")
+	structureSQL, err := p.reader.GetStructure()
+	if err != nil {
+		return errors.Wrap(err, "failed to get structure")
+	}
+
+	if err := p.DumpStructure(structureSQL); err != nil {
+		return errors.Wrap(err, "failed to dump structure")
+	}
+
+	log.Debug("Structure dumped")
+	return nil
+}
+
+func (p *sqlDumper) readAndDumpTables(done chan<- struct{}, configTables config.Tables) error {
+	// Get the tables
+	tables, err := p.reader.GetTables()
+	if err != nil {
+		return err
+	}
+
+	// Trigger pre dump tables
+	if adv, ok := p.SqlEngine.(SqlEngineAdvanced); ok {
+		if err := adv.PreDumpTables(tables); err != nil {
+			return err
+		}
+	}
+
+	// TODO make the amount on concurrent dumps configurable
+	semaphoreChan := make(chan struct{}, 10)
+
+	var wg sync.WaitGroup
+	wg.Add(len(tables))
+	for _, tbl := range tables {
+		semaphoreChan <- struct{}{}
+
+		var opts reader.ReadTableOpt
+
+		table, err := configTables.FindByName(tbl)
+		if err != nil {
+			log.WithError(err).WithField("table", tbl).Debug("no configuration found for table")
+		}
+
+		if table != nil {
+			opts = reader.ReadTableOpt{
+				Limit:         table.Filter.Limit,
+				Relationships: p.relationshipConfigToOptions(table.Relationships),
+			}
+		}
+
+		// Create read/write chanel
+		rowChan := make(chan database.Row)
+
+		go func(tableName string, rowChan <-chan database.Row) {
+			if err := p.DumpTable(tableName, rowChan); err != nil {
```

Comment:
Here I would:

```go
defer wg.Done()
defer func(semaphoreChan <-chan struct{}) { <-semaphoreChan }(semaphoreChan)

if err := p.DumpTable(tableName, rowChan); err != nil {
  log.WithError(err).WithField("table", tableName).Error("Failed to dump table")
}
```

Reply:
Hm we also need to `close(rowChan)`

Reply:
rowChan is closed in the reader https://github.com/hellofresh/klepto/pull/58/files#diff-f7a183c94b4fefc7a5450142e7ef915aR176. We probably should change something in that behavior but it works for this PR.

Reply:
Ah gotcha 👍